File: app/controllers/calculator_controller.rb
class CalculatorController < ApplicationController
  def index
  end

  def calculate
    bill_amount = params[:bill].to_f
    tip_percentage = params[:tip].to_f
    people = params[:people].to_i

    tip_amount = (bill_amount * tip_percentage / 100)
    total_bill = bill_amount + tip_amount
    amount_per_person = total_bill / people

    TipCalculation.create!(
      bill_amount: bill_amount,
      tip_percentage: tip_percentage,
      tip_amount: tip_amount,
      total_bill: total_bill,
      number_of_people: people,
      amount_per_person: amount_per_person
    )

    render json: {
      tip_amount: format('%.2f', tip_amount / people),
      total_per_person: format('%.2f', amount_per_person)
    }
  end
end

File: app/javascript/controllers/calculator_controller.js

import { Controller } from "@hotwired/stimulus"

export default class extends Controller {
  static targets = ["bill", "customTip", "people", "tipAmount", "totalAmount"]
  
  connect() {
    this.selectedTip = 0
  }

  selectTip(event) {
    this.selectedTip = parseInt(event.currentTarget.dataset.tip)
    this.calculate()
  }

  customTipInput(event) {
    this.selectedTip = parseFloat(event.currentTarget.value)
    this.calculate()
  }

  calculate() {
    const bill = parseFloat(this.billTarget.value) || 0
    const people = parseInt(this.peopleTarget.value) || 1
    const tipPercent = this.selectedTip

    if (bill === 0 || people === 0) return

    fetch('/calculate', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'X-CSRF-Token': document.querySelector('[name="csrf-token"]').content
      },
      body: JSON.stringify({ bill, tip: tipPercent, people })
    })
    .then(response => response.json())
    .then(data => {
      this.tipAmountTarget.textContent = `$${data.tip_amount}`
      this.totalAmountTarget.textContent = `$${data.total_per_person}`
    })
  }

  reset() {
    this.billTarget.value = ''
    this.customTipTarget.value = ''
    this.peopleTarget.value = ''
    this.tipAmountTarget.textContent = '$0.00'
    this.totalAmountTarget.textContent = '$0.00'
    this.selectedTip = 0
  }
}